### PR TITLE
Fixed validation of TRUSTED CA bundles

### PR DESCRIPTION
**Pull request by AnrDaemon**
`CaBundle::validateCaFile` uses function `openssl_x509_parse`, which can only parse regular certificate files, to parse trusted CA bundles (`BEGIN TRUSTED CERTIFICATE` format).

Since there's no alternative to parse such format directly, this little hack will let the parsing succeed.

OpenSSL extension itself don't have an issue with these bundles and handle host validation correctly.

### CODE REVIEW

**Comment by garas** [comment]
What is the source of `ca-bundle.trust.crt`? How to get updates, what is the maintainer?

**Comment by AnrDaemon** [comment]
Same source. I.e. Mozilla PKI. It is generated by update scripts in CentOS and Cygwin, for example.

**Comment by Seldaek** [comment]
I am not sure I see the point of adding this additional file. Removing all text between `-----END (TRUSTED)? CERTIFICATE-----` and `-----BEGIN (TRUSTED)? CERTIFICATE-----` should achieve the same for all files no? The point of the validation is to determine if the file is usable or not, which I think the current way achieves quite well as it doesn't seem we are having a problem. openssl_x509_parse can parse the cacert.pem we ship, even though I guess it only parses the first certificate in the file.

**Comment by AnrDaemon** [comment]
> I am not sure I see the point of adding this additional file. 

I'm not sure what file you are referring to. The `res/ca-bundle.trust.crt` ? I added it as an example resource for the tests. I've considering adding just part of it (first certificate), but seeing that the `cacert.pem` included in full, I've had a doubt.

> Removing all text between -----END (TRUSTED)? CERTIFICATE----- and -----BEGIN (TRUSTED)? CERTIFICATE----- should achieve the same for all files no?

That's how I hacked validation.

> The point of the validation is to determine if the file is usable or not, which I think the current way achieves quite well as it doesn't seem we are having a problem.

I do have a problem. I'm using private CA to sign certificates for my infrastructure, and I'm using `ca_bundle.trust.crt` as a source of trust for my system. It is configured in `php.ini`, and curl and PHP itself connect to my webservers just fine, but Composer fails to validate it and injecting its own CA bundle instead.

**Comment by Seldaek** [comment]
Ok then why is it failing to validate it? Is it because your file uses `BEGIN TRUSTED CERTIFICATE` and not `BEGIN CERTIFICATE`?

**Comment by AnrDaemon** [comment]
Exactly. `openssl_x509_parse` does not recognize certificates in trusted CA bundle. Not supposed to, I suppose.

P.S.
It is not "my file". It is a standard trusted CA file (certificates with additional trust record).

**Comment by Seldaek** [comment]
Ok, so fine by me but the .trust.crt file should be in tests/Fixtures folder or something, not in res/ as it's not something I want to spend time updating. And it should only contain one or two certs, that's enough to run tests.

File: src/CaBundle.php [hunk 1]
```diff
@@ -160,6 +160,7 @@ public static function validateCaFile($filename, LoggerInterface $logger = null)
         }
 
         $contents = file_get_contents($filename);
+        $contents = preg_replace("/^(\\-+(?:BEGIN|END))\\s+TRUSTED\\s+(CERTIFICATE\\-+)\$/m", '$1 $2', $contents);
```
Comment (Seldaek) [inline src/CaBundle.php:163]: Actually this could be moved right before the openssl_x509_parse call, as it's not really needed otherwise?

**Comment by Seldaek** [comment]
Thanks for the updates, just one last little nitpick then good to merge for me 👍

**Comment by AnrDaemon** [comment]
Amended that.

**Comment by Seldaek** [comment]
Thanks!

**Pull request merged**